src/node_i18n.cc

```diff
@@ -62,6 +72,306 @@ bool flag_icu_data_dir = false;
+  status = U_ZERO_ERROR;
+
+  uint32_t len = ucnv_fromUChars(to.conv, *buf, length,
+                                 source, length, &status);
+
+  if (U_SUCCESS(status)) {
+    MaybeLocal<Object> ret = AsBuffer(isolate, &buf, len);
+    if (!ret.IsEmpty())
+      return args.GetReturnValue().Set(ret.ToLocalChecked());
+  }
+
+  ThrowICUError(env, status, "Unable to transcode buffer");
+}
+
+// Converts from UTF-8 to UCS2
+// args[0] must be a buffer instance
+void Ucs2FromUtf8(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
+  SPREAD_BUFFER_ARG(args[0], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+  MaybeStackBuffer<UChar, kStorageSize> buf;
+  int32_t len;
+
+  u_strFromUTF8(*buf, kStorageSize, &len, ts_obj_data, ts_obj_length, &status);
+
+  MaybeLocal<Object> ret;
+  if (U_SUCCESS(status)) {
+    ret = AsBuffer(isolate, &buf, len * sizeof(**buf));
+  } else if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    u_strFromUTF8(*buf, len, &len, ts_obj_data, ts_obj_length, &status);
+    if (U_SUCCESS(status))
+      ret = AsBuffer(isolate, &buf, len * sizeof(*buf));
+  }
+
+  if (!ret.IsEmpty())
+    return args.GetReturnValue().Set(ret.ToLocalChecked());
+
+  ThrowICUError(env, status, "Unable to transcode buffer");
+}
+
+// Converts UCS2 into UTF-8
+// args[0] must be a buffer instance
+void Utf8FromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
+  SPREAD_BUFFER_ARG(args[0], ts_obj);
+
+  const size_t length = ts_obj_length >> 1;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  swapspace.AllocateSufficientStorage(length);
+  if (IsLittleEndian()) {
+    memcpy(*swapspace, ts_obj_data, ts_obj_length);
+  } else {
+    for (size_t n = 0, i = 0; i < length; n += 2, i += 1) {
+      const uint8_t hi = static_cast<uint8_t>(ts_obj_data[n + 0]);
+      const uint8_t lo = static_cast<uint8_t>(ts_obj_data[n + 1]);
+      swapspace[i] = (lo << 8) | hi;
+    }
+  }
+  source = *swapspace;
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<char, kStorageSize> buf;
+  int32_t len;
+
+  u_strToUTF8(*buf, kStorageSize, &len, source, length, &status);
+  MaybeLocal<Object> ret;
+  if (U_SUCCESS(status)) {
+    ret = AsBuffer(isolate, &buf, len);
+  } else if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    u_strToUTF8(*buf, len, &len, source, length, &status);
+    if (U_SUCCESS(status)) {
+      ret = Buffer::New(isolate, *buf, len);
+      buf.Release();
+    }
+  }
+
+  if (!ret.IsEmpty())
+    return args.GetReturnValue().Set(ret.ToLocalChecked());
+
+  ThrowICUError(env, status, "Unable to transcode buffer");
+}
```

Comment:
There are 50-100 lines of code in these two functions that are mostly the same.  Honestly, that's too much duplication to my liking.